packages/protocol/contracts/governance/Governance.sol

```diff
@@ -529,49 +545,36 @@ contract Governance is
   function vote(
     uint256 proposalId,
     uint256 index,
-    VoteValue value
+    Proposals.VoteValue value
   )
     external
     returns (bool)
   {
     address account = getAccountFromVoter(msg.sender);
     require(!isVotingFrozen(account));
     dequeueProposalsIfReady();
-    Proposal storage proposal = proposals[proposalId];
-    require(_proposalExists(proposal) && dequeued[index] == proposalId);
-    if (isDequeuedProposalExpired(proposalId)) {
-      deleteDequeuedProposal(proposalId, index);
+    Proposals.Proposal storage proposal = proposals[proposalId];
+    require(isDequeuedProposal(proposal, proposalId, index));
+    Proposals.Stage stage = proposal.getDequeuedStage(stageDurations);
+    if (isDequeuedProposalExpired(proposal, stage)) {
+      deleteDequeuedProposal(proposal, proposalId, index);
       return false;
     }
-    ProposalStage stage = _getDequeuedProposalStage(proposal.timestamp);
     Voter storage voter = voters[account];
     uint256 weight = getAccountWeight(account);
     require(
-      proposal.approved &&
-      stage == ProposalStage.Referendum &&
-      value != VoteValue.None &&
+      proposal.isApproved() &&
+      stage == Proposals.Stage.Referendum &&
+      value != Proposals.VoteValue.None &&
       weight > 0
     );
     VoteRecord storage voteRecord = voter.referendumVotes[index];
-    // If we've already voted on this proposal, subtract the previous vote.
-    if (voteRecord.proposalId == proposalId) {
-      if (voteRecord.value == VoteValue.Abstain) {
-        proposal.votes.abstain = proposal.votes.abstain.sub(weight);
-      } else if (voteRecord.value == VoteValue.Yes) {
-        proposal.votes.yes = proposal.votes.yes.sub(weight);
-      } else if (voteRecord.value == VoteValue.No) {
-        proposal.votes.no = proposal.votes.no.sub(weight);
-      }
-    }
-
-    // Add new vote.
-    if (value == VoteValue.Abstain) {
-      proposal.votes.abstain = proposal.votes.abstain.add(weight);
-    } else if (value == VoteValue.Yes) {
-      proposal.votes.yes = proposal.votes.yes.add(weight);
-    } else if (value == VoteValue.No) {
-      proposal.votes.no = proposal.votes.no.add(weight);
-    }
+    proposal.vote(
+      weight,
+      (voteRecord.proposalId == proposalId) ? voteRecord.value : Proposals.VoteValue.None,
```

Comment:
1) That was my understanding at first, but VoteRecord is actually mapped to by dequeued index, not by proposal index. That's why we check that the proposal id also matches.

2) I don't feel strongly about preserving the name of the library function `vote()`, especially since its signature is pretty different from the external `vote()` called in Governance. Does `updateVote()` capture what's happening well?